### PR TITLE
Move all unittest processing to the unittest target

### LINKED ISSUE #41: Remove `-unittest` from default `test/` targets flags and run unittests in `test/` via `make unittest` instead
Current proposed solution
=====================

* Remove `-unittest` when compiling integrations tests (`test/`) because they complicate debugging.

* When getting the files to run `unittest`s for the `make unittest` target, also consider the integration tests files (in `test/`), so integrations test there still ran at least in some context.

This is a good trade-off to have integration tests still running but keeping clean runs for integration tests (also avoid a lot of duplication when running `unittest` blocks, which now  will be ran only once via the `make unittest` target).

Original issue description
====================

Originally I have pushed for keeping the flag so that internal sanity of test suite can also be validated. However practice has proven that this is almost never used. At the same time, running all unittests of all modules has also proven to complicate debugging a lot.

I propose to publicly shame me and remove that flag by default.

### PR DESCRIPTION
**Pull request by leandro-lucarella-sociomantic**
This commit moves the compilation and running of unittests to the
`unittest` target(s) and remove the flags `-unittest -debug=UnitTest
-version=UnitTest` when building integrationtests.

Fixes #41.

### CODE REVIEW

**Comment by leandro-lucarella-sociomantic** [comment]
@mathias-baumann-sociomantic I would assign you but you didn't accept the invitation yet :)

**Comment by leandro-lucarella-sociomantic** [comment]
BTW, untested, please do test!

File: Makd.mak
Comment (mihails-strasuns-sociomantic) [inline Makd.mak]: This results in stuff like this in allunittests.d
```D
import .docker.ocean.test.unixlistener.main.d;
```
Maybe has something to do how `find_files` processes qualified paths?

**Review by mihails-strasuns-sociomantic — CHANGES_REQUESTED** [review]

File: Makd.mak
Comment (leandro-lucarella-sociomantic) [reply Makd.mak]: Mmm, weird. It shouldn't.

**Comment by mihails-strasuns-sociomantic** [comment]
Ping @leandro-lucarella-sociomantic

**Comment by leandro-lucarella-sociomantic** [comment]
I found the problem, but we have  another issue, the extra `main()`s.

For now for the program we get around this by just not building/running unittests in the `main.d` file, but for tests is normal that they only have a `main.d` so skipping those will mean no running unittests at all. I think the only sensible thing to do here is to stop filtering-out `main.d` files and just force projects to use something like:

```d
version (UnitTest) {}
else
int main(...)
{
    // ...
}
```

Ideas/opinions?

**Comment by leandro-lucarella-sociomantic** [comment]
@gavin-norman-sociomantic FYI, reading the last comment might be enough.

**Comment by leandro-lucarella-sociomantic** [comment]
If we go this route, I need to update this PR to fix docs/release notes to note:

* The project's `main.d` files will no longer be auto-filtered-out from unittests (if we'll require that for `integrationtest` I think it doesn't make sense to exclude any files from `make unittest`, it's unnatural.

* All `main()` files need to be versioned-out when they are compiled with `version=UnitTest`.

* New `$(INTEGRATIONTEST)` variable.

**Comment by mihails-strasuns-sociomantic** [comment]
I don't think we should do anything too smart here. This will come in a new makd major release - migration notes explaining the change and asking to adjust own code in appropriate manner would be enough.

**Comment by leandro-lucarella-sociomantic** [comment]
> I don't think we should do anything too smart here. This will come in a new makd major release - migration notes explaining the change and asking to adjust own code in appropriate manner would be enough.

This means you agree with the proposed changes?

**Comment by leandro-lucarella-sociomantic** [comment]
ping

**Comment by mihails-strasuns-sociomantic** [comment]
What do you mean by `INTEGRATIONTEST` variable?

File: Makd.mak
Comment (leandro-lucarella-sociomantic) [inline Makd.mak]: This

**Comment by leandro-lucarella-sociomantic** [comment]
Just a way to ease the path to using `integrationtest` instead in the future. Not really related to this but it was good to make it easier to find all the places where `test/` was used.

**Comment by mihails-strasuns-sociomantic** [comment]
I see. Yeah, I do agree with propose path in that case.

**Comment by leandro-lucarella-sociomantic** [comment]
OK, I will update with the docs before I forget what I've done :D

**Comment by leandro-lucarella-sociomantic** [comment]
> What do you mean by INTEGRATIONTEST variable?

While at this, I think I will backport this to v1.x.x, so projects can already try changing `test/` to `integrationtest/` to see if that really fixes our problems with D2 and see if it fixes #74.

If this is the case, we can change the default in v2.x.x more confidently!

**Comment by leandro-lucarella-sociomantic** [comment]
Updated! Please give this a close look before merging, it was a bit too long ago since last time I touched it :)

**Comment by leandro-lucarella-sociomantic** [comment]
`INTEGRATIONTEST` backported via #82 which also addresses the related #74.

**Comment by leandro-lucarella-sociomantic** [comment]
Added a commit to fix #74 (change the default location for integration tests).

**Comment by leandro-lucarella-sociomantic** [comment]
ping @mihails-strasuns-sociomantic

**Comment by mihails-strasuns-sociomantic** [comment]
Will re-test after #74 is merged

**Comment by leandro-lucarella-sociomantic** [comment]
You mean #82, right?

**Comment by mihails-strasuns-sociomantic** [comment]
Yep

**Comment by mihails-strasuns-sociomantic** [comment]
Rebase?

**Comment by leandro-lucarella-sociomantic** [comment]
Mmm, no? It seems to be at the current v2.x.x.

**Comment by leandro-lucarella-sociomantic** [comment]
Ah, maybe you mean merge, because the other changes were done to v1.x.x, I think I will merge after the release of v1, but I guess I could merge now if you prefer.

**Comment by leandro-lucarella-sociomantic** [comment]
OK, updated on top of the merge.

**Comment by leandro-lucarella-sociomantic** [comment]
Created a new commit with just the merge: https://github.com/sociomantic-tsunami/makd/pull/83

**Comment by leandro-lucarella-sociomantic** [comment]
So this PR can get less noisy after the merge is merged :D

**Comment by leandro-lucarella-sociomantic** [comment]
OK, now it's clean again.

**Review by mihails-strasuns-sociomantic — APPROVED** [review]
Looks good!

**Comment by leandro-lucarella-sociomantic** [comment]
:tada:

**Pull request merged**
